src/main/scala/mesosphere/marathon/util/LifeCycledCloseable.scala

```diff
@@ -0,0 +1,39 @@
+package mesosphere.marathon
+package util
+
+import java.io.Closeable
+import scala.util.Try
+
+trait LifeCycledCloseableLike[T <: Closeable] extends Closeable {
+  def closeable: T
+  def beforeClose(fn: () => Unit): Unit
+  def removeBeforeClose(fn: () => Unit): Unit
+}
+
+class LifeCycledCloseable[T <: Closeable](val closeable: T) extends LifeCycledCloseableLike[T] {
+  @volatile private[this] var closed: Boolean = false
```

Comment:
I'm pretty sure both variables doesn't need to be `volatile` since they are accessed within `synchronized` blocks.

Reply:
maybe `AtomicBoolean` will be a better option here?

Reply:
✅ 

Put the whole state in a single `AtomicReference`, instead :). This completely avoids the potential deadlock scenario (callings hooks inside of synchronized) and reduces any confusion about what is volatile.